src/logging/mod.rs

```diff
@@ -0,0 +1,162 @@
+/// Logging via slog
+///
+/// Provides a MozlogLogger with moz log fields per request
+ 
+use std::io;
+use std::ops::Deref;
+
+use failure::{err_msg};
+use rocket::{
+    Request, State,
+};
+use slog::{self, Drain, Record, Serializer, KV, Value, Key};
+use slog_async;
+use slog_mozlog_json::MozLogJson;
+use slog_term;
+
+use app_errors::Result;
+use settings::*;
+
+lazy_static! {
+    static ref LOGGER_NAME: String =
+        format!("{}-{}", env!("CARGO_PKG_NAME"), env!("CARGO_PKG_VERSION"));
+    static ref MSG_TYPE: String = format!("{}:log", env!("CARGO_PKG_NAME"));
+}
+
+#[derive(Clone)]
+struct RequestMozlogFields {
+    method: &'static str,
+    path: String,
+    remote: Option<String>,
+    agent: Option<String>,
+}
+
+impl RequestMozlogFields {
+    pub fn from_request(request: &Request) -> RequestMozlogFields {
+        let headers = request.headers();
+        RequestMozlogFields {
+            method: request.method().as_str(),
+            path: request.uri().to_string(),
+            agent: headers.get_one("User-Agent").map(&str::to_owned),
+            remote: headers
+                .get_one("X-Forwarded-For")
+                .map(&str::to_owned)
+                .or(request.remote().map(|addr| addr.ip().to_string())),
+        }
+    }
+}
+
+impl KV for RequestMozlogFields {
+    fn serialize(&self, _: &Record, serializer: &mut Serializer) -> slog::Result {
+        if let Some(ref agent) = self.agent {
+            serializer.emit_str("agent", agent)?;
+        }
+        if let Some(ref remote) = self.remote {
+            serializer.emit_str("remoteAddressChain", remote)?;
+        }
+        serializer.emit_str("path", &self.path)?;
+        serializer.emit_str("method", self.method)?;
+
+        Ok(())
+    }
+}
+
+impl Value for Settings {
+    fn serialize(&self, _: &Record, _: Key, serializer: &mut Serializer) -> slog::Result {
+        serializer.emit_str("auth_db_uri", &self.authdb.baseuri)?;
+        serializer.emit_str("provider", &self.provider)?;
+
+        if let Some(_) = self.aws.keys {
+            serializer.emit_str("aws_keys_access", "[hidden]")?;
+        } else {
+            serializer.emit_str("aws_keys_access", "[not set]")?;
```

Comment:
This is good, should we also include `keys.secret`?

Reply:
Since we are not going to show either keys nor access, i thought it was best to just ground them together and show `[hidden]` or `[not set]`. I guess this got a little bit confusing, because I called the property `aws_keys_access`. I'll just change it to `aws_keys`. Sounds better?